File: BOJ/[16234] 인구 이동/DAEGUE/16234.py
import sys
input = sys.stdin.readline

def dfs(start):
    sy, sx = start
    union[sy][sx] = num
    sum_list[num][0] += F_list[sy][sx]
    sum_list[num][1] += 1
    for i in range(4):
        ny = sy + dy[i]
        nx = sx + dx[i]
        if 0 <= ny < N and 0 <= nx < N:
            if not union[ny][nx]:
                if L <= abs(F_list[ny][nx] - F_list[sy][sx]) <= R:
                    dfs((ny, nx))

def move():
    for i in range(N):
        for j in range(N):
            F_list[i][j] = sum_list[union[i][j]][0] // sum_list[union[i][j]][1]

N, L, R = map(int, input().split())
F_list = []
for _ in range(N):
    row = list(map(int, input().split()))
    F_list.append(row)
count = 0
dx = [1, -1, 0 ,0]
dy = [0, 0, 1, -1]

while True:
    num = 1
    union = [[0] * N for _ in range(N)]
    sum_list = [[0, 0] for _ in range(N**2+1)]
    for i in range(N):
        for j in range(N):
            if not union[i][j]:
                dfs((i, j))
                num += 1
    if union[-1][-1] == N**2:
        break
    move()
    count += 1
print(count)

File: BOJ/[16234] 인구 이동/pinkishincoloragain/16234.js
const input = require("fs")
	.readFileSync("./dev/stdin")
	// .readFileSync("./input.txt")
	.toString()
	.trim()
	.split("\n");

const [N, L, R] = input
	.shift()
	.split(" ")
	.map(x => +x);

const map = [];

input.forEach(line => {
	map.push(line.split(" ").map(x => +x));
});

const move = [
	[0, 1],
	[1, 0],
	[0, -1],
	[-1, 0],
];

const dfs = (start, visited) => {
	const stack = [start];
	const union = [start];
	let population = map[start[0]][start[1]];
	visited[start[0]][start[1]] = true;
	while (stack.length) {
		const [x, y] = stack.pop();
		for (let k = 0; k < 4; k++) {
			const nx = x + move[k][0];
			const ny = y + move[k][1];

			if (nx < 0 || ny < 0 || N - 1 < nx || N - 1 < ny) continue;
			if (visited[nx][ny]) continue;

			const diff = Math.abs(map[x][y] - map[nx][ny]);

			if (L <= diff && diff <= R) {
				stack.push([nx, ny]);
				union.push([nx, ny]);
				population += map[nx][ny];
				visited[nx][ny] = true;
			}
		}
	}
	union.forEach(
		([x, y]) => (map[x][y] = Math.floor(population / union.length))
	);
	return union.length;
};

let cnt = 0;

while (true) {
	const visited = Array.from({ length: N }, () => Array(N).fill(false));
	let flag = false;
	for (let i = 0; i < N; i++) {
		for (let j = 0; j < N; j++) {
			if (visited[i][j]) continue;

			const unionCnt = dfs([i, j], visited);
			if (unionCnt > 1) flag = true;
		}
	}
	if (!flag) break;
	cnt++;
}

console.log(cnt);
